main/simulator.cpp

```diff
@@ -1 +1,78 @@
 #include "main/simulator.h"
+
+std::vector<std::unique_ptr<objects::EnergyCollector>>
+buildCollectors(const ModelInterface &model, int numCollectors) {
+
+  if (model.empty()) {
+    throw std::invalid_argument("Empty model");
+  }
+  if (numCollectors < 4) {
+    std::stringstream ss;
+    ss << "numCollectors: " << numCollectors << " is less then 4";
+    throw std::invalid_argument(ss.str());
+  }
+
+  // We distribute collectors along 4 quarter circles above the model, so
+  // |numCollectors| has to be divisible by 4. Or |numCollectors|-1 has to be
+  // divisible by 4 (if it is odd), in which case there is one more collector
+  // placed right above the model.
+  if (numCollectors % 4 != 0 && (numCollectors - 1) % 4 != 0) {
+    std::stringstream ss;
+    ss << "numCollectors or numCollectors-1 has to be divisible by 4, got "
+       << "numCollectors = " << numCollectors;
+    throw std::invalid_argument(ss.str());
+  }
+
+  // Minimum distance from Vec3(0, 0, 0) that each collector must meet.
+  // Value of the minDistance can't be less then 4 [m]
+  float minDistance =
+      4 * std::max<float>(
+              {model.height(), model.sideSize(),
+               /*makes sure that minDistance is at least equal to 4 */ 1.0f});
+
+  // For even number of ray collectors, in order to equally cover two simulation
```

Comment:
>  // For even number of ray collectors, in order to equally cover two simulation
>  // half-circle (half-circle that cross each collectors origin), we need to

1. half-circle**s** (plural)
2. `half-circle that cross each collectors origin`, this seems backwards: you are putting the collectors there, not the other way around. Remove.
3. there is no info about odd.

>  // calculate angle between Vec3(0, 0, 0) and 2 points where energy collector
>  // radius will cross simulation half-circle.

Not sure if I follow. What do you mean by "radius will cross circle"?

> // Calculation will varies depending on if number of collectors is even or not.

This is obvious right? Remove.

In general you are not explaining where the number of collectors in the denominator comes from, which was the key issue in the last code.

How about something like this:

```
We are placing collectors on two half-circles, such that distance between each collector on a given circle is equal to their radius.

For even number of collectors, on each half-circle we place N/2 collectors in total, with 2 collectors at the ground level. This divides the PI into N/2-1 segments.

For odd number of collectors, we do the same and place 1 more collector at the top, so we divide PI into (N-1)/2 segments.

The radius of one collector (distance between collectors) can be then calculated with Law of cosines, as R * sqrt(2 - 2 * cos(angle)).
```
Not sure if you need to change the formula or not.

Reply:
done. I refractored your foruma a ltlle bit, to much the code a littble bit better. I hope its better now ;)